File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/landing/SqlServerQuery.scala
package dev.typr.foundations.docs.landing
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*
import dev.typr.foundationssc.data.*

import java.sql.Connection

@SuppressWarnings(Array("unused"))
object SqlServerQuery:
  case class OrderRow(id: Int, name: String, price: BigDecimal)
  val orderRowCodec: RowCodec[OrderRow] = null // placeholder
  val maxPrice: Option[BigDecimal] = None
  val conn: Connection = null // placeholder

  //start
  // Build small reusable filters - SQL Server example
  val nvarchar = SqlServerTypes.nvarchar
  val decimal = SqlServerTypes.decimal

  def byName(name: String): Fragment =
    sql"name LIKE ${nvarchar(name)}"

  def cheaperThan(max: BigDecimal): Fragment =
    sql"price < ${decimal(max)}"

  // Compose dynamically
  val filters: List[Fragment] =
    List(
      Some(byName("%widget%")),
      maxPrice.map(cheaperThan)
    ).flatten

  val orders: List[OrderRow] =
    sql"SELECT * FROM orders ${Fragment.whereAnd(filters)}"
      .query(orderRowCodec.all())
      .run(conn)
  //stop

File: documentation-examples-kotlin/src/kotlin/dev/typr/foundations/docs/landing/SqlServerQuery.kt
package dev.typr.foundations.docs.landing

import dev.typr.foundationskt.*
import dev.typr.foundationskt.data.*
import java.math.BigDecimal
import java.sql.Connection

@Suppress("unused")
class SqlServerQuery {
    data class OrderRow(val id: Int, val name: String, val price: BigDecimal)
    val orderRowCodec: RowCodec<OrderRow> =
        RowCodec.builder<OrderRow>()
            .field(SqlServerTypes.int_, OrderRow::id)
            .field(SqlServerTypes.nvarchar, OrderRow::name)
            .field(SqlServerTypes.decimal, OrderRow::price)
            .build(::OrderRow)
    val maxPrice: BigDecimal? = null
    lateinit var conn: Connection

    //start
    // Build small reusable filters - SQL Server example
    fun byName(name: String): Fragment =
        sql { "name LIKE ${SqlServerTypes.nvarchar(name)}" }

    fun cheaperThan(max: BigDecimal): Fragment =
        sql { "price < ${SqlServerTypes.decimal(max)}" }

    // Compose dynamically - only include the filters that are present
    val filters: List<Fragment> =
        listOfNotNull(
            byName("%widget%"),
            maxPrice?.let { cheaperThan(it) }
        )

    val orders: List<OrderRow> =
        sql { "SELECT * FROM orders ${Fragment.whereAnd(filters)}" }
            .query(orderRowCodec.all())
            .run(conn)
    //stop
}
